[CSS] Mark selector as implicit to ignore them in specificity calculation

mdubet opened the pull request:
#### dedf708653ff64d0743e6be773311fec317cd601
<pre>
[CSS] Mark selector as implicit to ignore them in specificity calculation
<a href="https://bugs.webkit.org/show_bug.cgi?id=265992">https://bugs.webkit.org/show_bug.cgi?id=265992</a>
<a href="https://rdar.apple.com/119307433">rdar://119307433</a>

Reviewed by Antti Koivisto.

We sometimes insert non author written selectors in the selector list
(for cases like `:has(&gt; foo)` or scoped or nested selectors).
We should not count those implicit selectors in the specificity calculation.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-specificity-expected.txt:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::SelectorSpecificity::specificityTuple const):
(WebCore::SelectorSpecificity::debugDescription const):
(WebCore::operator&lt;&lt;):
(WebCore::selectorSpecificity):
(WebCore::maxSpecificity):
(WebCore::simpleSelectorSpecificity):
(WebCore::CSSSelector::computeSpecificityTuple const):
(WebCore::CSSSelector::CSSSelector):
* Source/WebCore/css/CSSSelector.h:
(WebCore::CSSSelector::setImplicit):
(WebCore::CSSSelector::isImplicit const):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeStyleRule):

Canonical link: <a href="https://commits.webkit.org/271675@main">https://commits.webkit.org/271675@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/20373b6188dab7bbc1787453f46962a483303ec1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29204 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7873 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30542 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31792 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26564 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29801 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/10015 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/5178 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26568 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29477 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6538 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/25027 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5636 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5779 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/26064 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/33129 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26675 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26497 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/32005 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5744 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3926 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29791 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7422 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6968 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/6257 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/6267 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/CSSSelector.h:
```diff
@@ -336,6 +336,9 @@ struct PossiblyQuotedIdentifier {
         void setForPage() { m_isForPage = true; }
 
+        void setImplicit() { m_isImplicit = true; }
+        bool isImplicit() const { return m_isImplicit; }
```
It would be good to add a comment explaining what 'implicit' means here.

mdubet committed:
[CSS] Mark selector as implicit to ignore them in specificity calculation
https://bugs.webkit.org/show_bug.cgi?id=265992
rdar://119307433

Reviewed by Antti Koivisto.

We sometimes insert non author written selectors in the selector list
(for cases like `:has(> foo)` or scoped or nested selectors).
We should not count those implicit selectors in the specificity calculation.

* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/scope-specificity-expected.txt:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::SelectorSpecificity::specificityTuple const):
(WebCore::SelectorSpecificity::debugDescription const):
(WebCore::operator<<):
(WebCore::selectorSpecificity):
(WebCore::maxSpecificity):
(WebCore::simpleSelectorSpecificity):
(WebCore::CSSSelector::computeSpecificityTuple const):
(WebCore::CSSSelector::CSSSelector):
* Source/WebCore/css/CSSSelector.h:
(WebCore::CSSSelector::setImplicit):
(WebCore::CSSSelector::isImplicit const):
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeStyleRule):

Canonical link: https://commits.webkit.org/271675@main
